File: cc_lazer.py
from collections import namedtuple

MAX_N = 100000 + 7

class FenwickTree:
    def __init__(self, n):
        self.t = [0 for _ in range(n+1)]
        self.local_n = None

    def sum(self, index):
        res = 0
        # print("Summing...")
        while index > 0:
            # print(f"index: {index}")
            res += self.t[index]
            index -= (index & (-index))
        return res

    def update(self, index, value):
        # print(f"Updating tree... for index: {index} and value {value}")
        while index < self.local_n:
            # print(f"index: {index}")
            self.t[index] += value
            index += (index & (-index))

t = int(input())


q_vec = [[0, 0, 0] for _ in range(MAX_N)]
res = [0 for _ in range(MAX_N)]
ft = FenwickTree(MAX_N)

for _ in range(t):
    nq = input().split(" ")
    n = int(nq[0])
    q = int(nq[1])

    ft.local_n = n + 1
    a_vec = [int(a) for a in input().split(" ")]

    for i in range(q):
        x1x2y = input().split(" ")

        q_vec[i][0] = int(x1x2y[0])
        q_vec[i][1] = int(x1x2y[1])
        q_vec[i][2] = int(x1x2y[2])

    events = []
    # Evt = namedtuple("Evt", ["y", "t", "i"])
    for i in range(n-1):
        e_low = (min(a_vec[i], a_vec[i+1]), 1, i)
        e_high = (max(a_vec[i], a_vec[i+1]), 3, i)  
        
        events.append(e_low)
        events.append(e_high)
    
    for i in range(q):
        e_lazer = (q_vec[i][2], 2, i)
        events.append(e_lazer)

    events.sort(key=lambda x: (x[0], x[1]))
    # print(events)

    for e in events:
        # print(f"e: {e}")
        if e[1] == 1:
            ft.update(e[2]+1, 1)
            # print(ft.t)
        elif e[1] == 2:
            # print(f"q_vec[e[2]][1]: {q_vec[e[2]-1][1]} q_vec[e[2]][0]: {q_vec[e[2]-1][0]}")
            res[e[2]] = ft.sum(q_vec[e[2]][1]-1) - ft.sum(q_vec[e[2]][0]-1)
        elif e[1] == 3:
            ft.update(e[2]+1, -1)
            # print(ft.t)

    for i in range(q):
        print(res[i])

File: cc_lazer.cpp
#include <bits/stdc++.h> 

using namespace std;
typedef long long int lli;
typedef unsigned long long int ulli;
typedef long double ld;

void left_just(string &s, int n, char c = ' ') {
    while(s.length() < n) {
        s = c + s;
    }
}

template<typename T> void print_vector(vector<T> &vec) {

    int n = vec.size();
    for(int i = 0; i < n; i++) {
        if(i == n - 1)
            cout << vec[i];
        else
            cout << vec[i] << " ";
    }
    cout << "\n";
}

template<typename T> void print_matrix(vector<vector<T>> &mat) {
    for(int i = 0; i < mat.size(); i++) {
        print_vector(mat[i]);
    }
}


class Event {
    public:
        Event() {y = 0; t = 0; i = 0;}
        Event(int iy, int it, int ii) {y = iy; t = it; i = ii;}
        int y;
        int t;
        int i;
};


int fenwick_sum(int index, vector<int> &fenwick_bit) {
    int res = 0;
    while(index > 0) {
        res += fenwick_bit[index];
        index -= (index & (-index));
    }
    return res;
}

void fenwick_update(int index, int value, vector<int> &fenwick_bit) {
    while(index < fenwick_bit.size()) {
        fenwick_bit[index] += value;
        index += (index & (-index));
    }
}


int main() {

    ios_base::sync_with_stdio(0);
    cin.tie(0);

    int T;
    cin >> T;

    for(int t = 0; t < T; t++) {

        int N, Q;
        cin >> N >> Q;

        vector<int> a_vec(N, 0);
        const int N_ROWS = 3;
        vector<vector<int>> q_vec(Q, vector<int>(N_ROWS, 0));

        for(int n = 0; n < N; n++) {
            cin >> a_vec[n];
        }

        for(int q = 0; q < Q; q++) {
            int x1, x2, y;
            cin >> x1 >> x2 >> y;
            q_vec[q][0] = x1;
            q_vec[q][1] = x2;
            q_vec[q][2] = y;
        }


        vector<Event> events;
        events.reserve(2*(N-1) + Q);
        for(int n = 0; n < N - 1; n++) {
            Event e_low = Event(min(a_vec[n], a_vec[n+1]), 1, n);
            Event e_high = Event(max(a_vec[n], a_vec[n+1]), 3, n);
        
            events.push_back(e_low);
            events.push_back(e_high);
        }

        for(int q = 0; q < Q; q++) {
            Event e_lazer = Event(q_vec[q][2], 2, q);
            events.push_back(e_lazer);
        }

        auto func = [](Event &e1, Event &e2){
            if(e1.y < e2.y) { 
                return true;
            } else if(e1.y == e2.y) {
                if(e1.t <= e2.t) {
                    return true;
                } else {
                    return false;
                }
            }
            return false;
        };
        sort(events.begin(), events.end(), func);

        // for(const auto & e : events) {
        //     cout << "e: " << e.y << " " << e.t << " " << e.i << "\n";
        // }

        vector<int> fenwick_bit(N, 0);
        vector<int> res(Q, 0);
        for(int i = 0; i < events.size(); i++) {
            if(events[i].t == 1) {
                fenwick_update(events[i].i+1, 1, fenwick_bit);
            } else if(events[i].t == 2) {
                res[events[i].i] = fenwick_sum(q_vec[events[i].i][1]-1, fenwick_bit) - fenwick_sum(q_vec[events[i].i][0]-1, fenwick_bit);
                // cout << "fenwick_sum(q_vec[events[i].i][1]-1, fenwick_bit): " << fenwick_sum(q_vec[events[i].i][1]-1, fenwick_bit) << "\n";
            } else if(events[i].t == 3) {
                fenwick_update(events[i].i+1, -1, fenwick_bit);
            }
        }

        for(int i = 0; i < res.size(); i++) {
            cout << res[i] << "\n";
        }



    }

}